pkg/aggregator/no_aggregation_stream_worker.go

```diff
@@ -0,0 +1,215 @@
+//   * listens for a stop signal
+//   * listens for a flush signal
+// This is not ideal since the serializer should automatically takes the decision when to flush payloads to
+// the serializer but that's not how it works today, see noAggregationStreamWorker comment.
+func (w *noAggregationStreamWorker) run() {
+	log.Debugf("Starting streaming routine for the no-aggregation pipeline")
+
+	ticker := time.NewTicker(time.Second * 2)
+	defer ticker.Stop()
+	lastStream := time.Now().Add(time.Hour * 24)
+
+	logPayloads := config.Datadog.GetBool("log_payloads")
+	w.seriesSink, w.sketchesSink = createIterableMetrics(w.flushConfig, w.serializer, logPayloads, false)
+
+	stopped := false
+	var flushBlockChan chan struct{}
+	var stopBlockChan chan struct{}
+
+	for !stopped {
+		start := time.Now()
+		serializedSamples := 0
+
+		metrics.Serialize(
+			w.seriesSink,
+			w.sketchesSink,
+			func(seriesSink metrics.SerieSink, sketchesSink metrics.SketchesSink) {
+			mainloop:
+				for {
+					select {
+					// stop signal
+					case trigger := <-w.stopChan:
+						stopped = true
+						stopBlockChan = trigger.blockChan
+						break mainloop
+
+					// ticker regularly producing a flush signal if necessary
+					case <-ticker.C:
+						n := time.Now()
+						if lastStream.Before(n.Add(-time.Second*1)) && serializedSamples > 0 {
+							log.Debug("noAggregationStreamWorker: triggering an automatic payloads flush to the forwarder (no traffic since 1s)")
+							go w.flush(true)
+							lastStream = n.Add(time.Hour * 24)
+						}
+
+					// flush signal
+					case trigger := <-w.flushChan:
+						flushBlockChan = trigger.blockChan
+						break mainloop
+
+					// receiving samples
+					case samples := <-w.samplesChan:
+						log.Debugf("Streaming %d metrics from the no-aggregation pipeline", len(samples))
+						for _, sample := range samples {
+							// enrich metric sample tags
+							sample.GetTags(w.taggerBuffer, w.metricBuffer)
+							w.metricBuffer.AppendHashlessAccumulator(w.taggerBuffer)
+
+							// turns this metric sample into a serie
+							var serie metrics.Serie
+							serie.Name = sample.Name
+							serie.Points = []metrics.Point{{Ts: sample.Timestamp, Value: sample.Value}}
+							serie.Tags = tagset.CompositeTagsFromSlice(w.metricBuffer.Copy())
+							serie.Host = sample.Host
+							serie.Interval = 0 // unused when late
+							w.seriesSink.Append(&serie)
+
+							w.taggerBuffer.Reset()
+							w.metricBuffer.Reset()
+						}
+
+						lastStream = time.Now()
+
+						serializedSamples += len(samples)
+						if serializedSamples > w.maxMetricsPerPayload {
+							go w.flush(true)
+						}
+					}
+				}
+			}, func(serieSource metrics.SerieSource) {
+				sendIterableSeries(w.serializer, start, serieSource)
+				// the flush trigger may have set this flushBlockChan to a channel on which
+				// we need to send a signal to indicate the end of the flush.
+				if flushBlockChan != nil {
+					flushBlockChan <- struct{}{}
```

Comment:
Would it be better to `close` the channel here instead of sending a signal and setting it to `nil`?

Reply:
I still have to set it to nil but you're right, would be better to close it I think.